### DISCUSSION TITLE
Add new host function to check the number of keys on the account

### DISCUSSION THREAD

**Discussion by frol** [near/nearcore#5738] [Ideas] [answered]
This is useful when you want to check if the contract account is locked (no keys are there)

I have initially considered some API that would allow to read all the keys in the contract context, but it seems quite trouble-some from the encoding point of view. A simple number or even a bool ("are there any keys on the account?") would already help to solve the security challenge that is is front of https://github.com/galactic3/name_hub project, which needs to ensure that nobody else can control the account but the trusted contract deployed to it.

There is a more high-level problem, that I described in [#5740](https://github.com/near/NEPs/discussions/294)

cc @matklad @evgenykuzyakov @mikedotexe @amgando @nearmax @bowenwang1996

**Comment by matklad** [comment]
Adding `fn acces_key_count() -> Result<u64>` seems good to me. If the need arises, we can layer "read all keys" on top of it later (see `fn promise_results_count(&mut self) -> Result<u64>` and `fn promise_result(&mut self, result_idx: u64, register_id: u64) -> Result<u64>`).

**Reply by frol** [reply]
I feel it can return `u64` (I don't see how it can fail), but other than that it is exactly what I am looking for. We might want to have two counts: the number of any access keys, and the number of full access keys.

**Reply by matklad** [reply]
>(I don't see how it can fail)

It can run out of gas 

EDIT: I am being confusing here -- internal impl would return a `Result`, as all host functions do. For the contracts, the signature would be `(import "env" "access_key_count" (func (result i64)))` and it'll panic in case of an error.

**Reply by frol** [reply]
I was under impression that running out of gas won't get control back to Wasm code anyway :thinking: 

Looking at [`signer_account_pk() -> PublicKey`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.signer_account_pk.html) I assumed there should not be the `Result` on the API layer while some things might indeed go wrong on the ABI layer which will cause panic.

**Comment by bowenwang1996** [comment, marked as answer]
Duplicate of https://github.com/near/NEPs/issues/79
